### PR TITLE
Feature/add pagination and streaming support

### PR DESCRIPTION
**Pull request by jdwil**
#217 This PR adds streaming support as well as automatic pagination. Both require a Doctrine QueryBuilder to be returned as the View data, so Doctrine DBAL is required to use the feature, but no dependency is added to rest-bundle.

If a QueryBuilder is returned in the View, and the limit and offset parameters are in the ParamFetcher (both keys are configurable in View) then it will automatically paginate the results and let rest-bundle deliver the content normally. It also automatically sets the "Content-Range" header in the response for user by front-end pagers.

If no pagination parameters are present, then the results of from the QueryBuilder will be streamed. When being streamed, the rows are fetched one at a time, encoded using the encoder or a callback set in View, and pushed to stdout. There is also a header and footer callback for more customizable output.

### CODE REVIEW

File: EventListener/ViewResponseListener.php [hunk 1]
```diff
@@ -96,6 +98,17 @@ public function onKernelView(GetResponseForControllerResultEvent $event)
         if (null === $view->getFormat()) {
             $view->setFormat($request->getRequestFormat());
         }
+        
+        $range = null;
+
+        if (class_exists("Doctrine\DBAL\Query\QueryBuilder")
```
Comment (Ener-Getick) [inline EventListener/ViewResponseListener.php]: The ``class_exists`` is useless here

**Comment by xabbuh** [comment]
I am not convinced that this feature belongs to the bundle. If it is currently not possible to easily implement this in userland code, I think we should rather think about exposing clean extension points to make it possible.

**Comment by jdwilliams19** [comment]
@xabbuh That's a fair point. I can refactor this to take a custom interface rather than a Doctrine QueryBuilder. Would it be okay to add an implementation of the interace for Doctrine to rest-bundle? Others could add implementations for Propel and other dbal's...

**Comment by xabbuh** [comment]
I more thought of an event that we could dispatch here which would allow to modify the view before it was transformed into a response.

**Comment by jdwilliams19** [comment]
Roger that. I've got a new version using a StreamableInterface and PageableInterface. I'll try to get the tests cleaned up this week and push it up. Then we can talk about the event idea... not sure I completely follow.

**Comment by jdwilliams19** [comment]
So, this is what the Pageable/Streamable interfaces look like. I did make a vendor-specific Doctrine dao in the DBAL code, which I've found useful. Before I write more tests, can I get some feedback on this approach? I've been using this code in a development project for a few weeks and it makes hooking up to a data table front end, like Bootstrap Table or jQuery DataTables, quite easy.

**Comment by Ener-Getick** [comment]
Sorry for the late answer.. I think we should at first implement @xabbuh proposition and then eventually include new classes to easily manage pagination and streaming even if i'm not sure this belongs to this bundle.

**Comment by jdwil** [comment]
@Ener-Getick @xabbuh Roger that. Thanks for the feedback. I'll look into refactoring this as I have time and creating a separate bundle.

**Comment by hipio** [comment]
Any update on this? A REST API should have pagination IMO.

**Comment by jdwil** [comment]
@hipio Sorry, I haven't had any time to work on this. I've been using my FOS Rest Bundle fork in my current side-project, which has the functionality described here. I don't want to roll it into production that way so I will circle back to this at some point.

**Comment by hipio** [comment]
@jdwil Ah, I see, no worries, just wondering as I'm having to add pagination as well. Using a fork isn't something I want to do though so I guess I'll have to wait or implement something myself without touching core Symfony files.

**Comment by c410-f3r** [comment]
Any news about this feature?

**Comment by vasilvestre** [comment]
Any news ? Fix checks ? Or already up ?

**Comment by Neirda24** [comment]
Any news ? Seems to me that any good API should have a pagination support. Keep it up ;)

**Comment by Tobion** [comment]
Closing as outdated.

**Pull request closed without merging**